packages/npm-dependencies-validation/src/index.ts

```diff
@@ -0,0 +1,68 @@
+import { spawnCommand } from "./utils/npmUtil";
+import { resolve, dirname } from "path";
+import { isManagedByNpm6 } from "./utils/packageJsonUtil";
+import {
+  DependencyIssue,
+  IssueType,
+  NpmLsDependency,
+  VscodeFsUri,
+} from "./types";
+
+const LS_DEPS = ["ls", "--depth=0"]; // dependencies and extraneous modules
+const LS_DEV_DEPS = [...LS_DEPS, "--dev"]; // devDependencies
+
+async function getDepsStatus(
+  packageJsonPath: string,
+  devDeps: boolean
+): Promise<any> {
+  // try to use DependencyType ???
+  return spawnCommand(
+    devDeps ? [...LS_DEV_DEPS] : [...LS_DEPS],
+    resolve(dirname(packageJsonPath))
+  );
+}
+
+function getProblemType(dependency: NpmLsDependency): IssueType | undefined {
+  if (dependency.missing) return "missing";
+  if (dependency.invalid) return "invalid";
+  if (dependency.extraneous) return "extraneous";
+}
+
+async function getIssues(
+  path: string,
+  devDeps = false
+): Promise<DependencyIssue[]> {
+  const { dependencies } = await getDepsStatus(path, devDeps);
+
+  const problematicDeps: DependencyIssue[] = [];
+  for (const name in dependencies) {
```

Comment:
seems like you are combining a **map** and **filter**.
try to do it in **two steps** using functional logic (lodash).

1. filter to only keep the deps that have issues
2. map those problemDeps to `NPMDepsIssues`

Basically when reading the code one is less interested in **how** you do it (imperative style).
but **what** you are trying to acomplish.

Reply:
dependencies is not an array it is an object. Like this:
"dependencies": {
    "typescript": {
      "required": "4.4.4",
      "missing": true
    }
  }